### PR TITLE
HPCC-12862 Need to support nothor and noroxie etc. as option on //version

### PR DESCRIPTION
**Pull request by AttilaVamos**
Add code to process exclusion(s) defined in //version tag

Update README.rst

Signed-off-by: Attila Vamos attila.vamos@gmail.com

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@ghalliday @richardkchapman please check and merge.

File: testing/regress/hpcc/regression/suite.py [hunk 1]
```diff
@@ -126,8 +126,17 @@ def addFileToSuite(self, eclfile):
             versions = eclfile.getVersions()
             versionId = 1
             for version in versions:
-                files.append({'basename':basename, 'version':version,  'id':versionId })
-                versionId += 1
+                if 'no'+self.name in version:
```
Comment (ghalliday) [inline testing/regress/hpcc/regression/suite.py]: Is this testing the cluster name, or the type of the cluster?

File: testing/regress/hpcc/regression/suite.py [hunk 2]
```diff
@@ -126,8 +126,17 @@ def addFileToSuite(self, eclfile):
...
+                    # Exclude it from this target
+                    pass
+                else:
+                    # Remove exclusion key(s) from version string
+                    config = getConfig();
```
Comment (ghalliday) [inline testing/regress/hpcc/regression/suite.py]: I don't think it should be looping through clusters here - the tag is a platform, so should only be "nothor","nohthor","noroxie", and I would ignore legacy "nothorlcr" (as lonbg as nothor is matched).

File: testing/regress/hpcc/regression/suite.py (hunk 1 above)
Comment (AttilaVamos) [inline testing/regress/hpcc/regression/suite.py]: It is testing the target name.

File: testing/regress/hpcc/regression/suite.py (hunk 2 above)
Comment (AttilaVamos) [inline testing/regress/hpcc/regression/suite.py]: This code as the whole Regression Engine uses the target/cluster defined in the Clusters array in the config file (ecl-test.json). Actually it is 
       "Clusters": [
            "hthor",
            "thor",
            "roxie"
        ],

Comment (ghalliday) [inline testing/regress/hpcc/regression/suite.py]: What does the list of clusters mean?  Is it the platform type (e.g., hthor/thor/roxie) or is it the names of the clusters - e.g., could be thor1 mythor etc?

Comment (AttilaVamos) [inline testing/regress/hpcc/regression/suite.py]: It is the platform type.

Comment (ghalliday) [inline testing/regress/hpcc/regression/suite.py]: I don't think it is - it looks like two concepts have got confused.

For example I have two hthor clusters on my machine - hthor and hthor2

If I do not have hthor2 in the cluster list I get the following error:

```
gavin:~/dev/hpcc/testing/regress$ ./ecl-test query 'bcd1.ecl' -t=hthor2
[Failure] 1. Unknown target cluster:'hthor2'!
[Error] Error (4000): Unknown cluster! 
^CTraceback (most recent call last):
  File "./ecl-test", line 274, in <module>
    regressMain.main()
  File "./ecl-test", line 269, in main
    self.regress.StopTimeoutThread()
 ```

If I add //nohthor to a test and run 
```./ecl-test query test.ecl -t=hthor```
then it isn't executed.

But if I run 

If I add //nohthor to a test and run 
```./ecl-test query test.ecl -t=hthor```
then it is.

That isn't correct.  The cluster is the name that is supplied to the -t=xxx.  The value used in //nohthor etc. is the platform type (which probably needs to be derived from the environment, or added to the cluster information).

Comment (AttilaVamos) [inline testing/regress/hpcc/regression/suite.py]: The lastes version of engine should display an error message for the second attempt:
./ecl-test query test.ecl -t=hthor2
if the hthor2 doesn't exists in Clusters array of config file and halt. (It does it for me if I try to run that command on my system)
If you have this cluster in your config, then you should use //nohthor2 in your file to exclude.

Or if you want to //nothor tag exclude the file from all clusters which name starts with hthor then I think, it is a different issue.

Comment (ghalliday) [inline testing/regress/hpcc/regression/suite.py]: Adding nothor2 to the clusters did allow me to run it.

However //nothor should mean do no run this on the thor platform, not do not run this on the thor cluster.

There should not be any reference to cluster names in any of the files.

**Comment by ghalliday** [comment]
@richardkchapman we could either merge this as-is and open another jira to fix nothor processing to use the platform in all places, or wait until this is fixed.

File: testing/regress/hpcc/regression/suite.py (hunk 2 above)
Comment (AttilaVamos) [inline testing/regress/hpcc/regression/suite.py]: (Above I mistyped the name, so the correct sentence is:” Or if you want to //nohthor tag excludes the file from all clusters which name starts with hthor then I think, it is a different issue.)
It wasn't problem in the last 1 - 1.5 year and we have couple of hundred official test results with this approach. Maybe the Clusters name (of the array in the config) is not accurate, we can change it easily. 
It is possible the terms (target, platform, cluster , etc.) are a little bit foggy and/or only in mind defined and/or mismatched/misused. We can clarify them easily too. 
BTW Actually the 

//version ... ,no<hthor|thor|roxie>

works in same way as the 

//no<hthor|thor|roxie>

tag does.  Therefore any or more no<hthor|thor|roxie> is defined in the //version tag then the

 ./ecl-test setup|run|query -t <hthor|thor|roxie> …

command doesn’t execute that version with 

ecl run –t <hthor|thor|roxie> …

command on the defined target.

If this approach is wrong then please raise a JIRA for this and specify how it should work.

**Comment by ghalliday** [comment]
Merging as-is because it will help clean up the output of the regression test suite, but opened https://track.hpccsystems.com/browse/HPCC-12891 to fix the way the tags are handled

**Comment by ghalliday** [comment]
@AttilaVamos I merged the change and then reverted it because the pull request was targetting the wrong branch.  Needs to be 5.2

**Pull request merged**